### DISCUSSION TITLE
LDM Notes for June 5th, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7309] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-06-05.md

## Agenda

- Collection literals

**Comment by orthoxerox** [comment]
> Further, no language that we can find that has dictionary literals differentiates them from collection literals with an extra or different sigil, so there's no prior art to draw on.

ECMAScript?

**Reply by 333fred** [reply]
I wouldn't really call that analogous to C# dictionaries: that's more akin to anonymous object creations.

**Reply by CyrusNajmabadi** [reply]
Yup. Js/es doesn't have dictionary literals (or 'map' in es parlance).  Rather, It has object literals.

**Reply by HaloFour** [reply]
Depending on who you ask, JS also doesn't have objects, it only has dictionaries.  😅

**Reply by qrli** [reply]
Doesn't Python use `{ }`?

**Reply by qrli** [reply]
And Ruby `{ "one" => 1 }` vs `[1, 2]`.
And PowerShell `@{ "one" = 1 }` vs `@(1, 2)`

**Reply by 333fred** [reply]
To be clearer, because I think you might be misinterpreting what my notes meant, we couldn't find a language that used a sigil _in addition to_ the list syntax. Ie, no precedent for the additional modification on the `[]` we discussed in the notes.

**Reply by qrli** [reply]
@333fred Thanks for the explanation. But that explanation brings another confusion:
You listed 3 possible options, with the 3rd being simple `{ }`, which is not using a sigil _in addition_. Then the notes stated "so there's no prior art to draw on. Given this, we don't think there needs to be a separate syntax for dictionary literals". How so? Why does it veto the 3rd option altogether?

**Reply by 333fred** [reply]
Because we simply didn't like it.

**Reply by qrli** [reply]
I think the syntax affects a bit the perceived behavior on order & duplicate elements.
- When using a different syntax for dictionary, it is a dictionary in and out. It does not preserve order nor duplicate elements. 
- When using the same syntax for dictionary as collection/sequence, it feels more like a collection/sequence of key-value pairs, thus preserving order & duplicate elements, until it is converted to a dictionary.

```cs
foreach (var kv in ["un" : 1, "deux" : 2, "trois" : 3, "un" : 1])
{
    Console.WriteLine(kv);
}
```
Of course, this is minor. It is fine to behave whatever you define.

**Reply by CyrusNajmabadi** [reply]
@qrli We've considered and discussed this.  As mentioned, it's not palatable for us.  This is esp. true as no good syntax has even been presented that fits cleanly into this space.  In our discussions, we've left the door open if someone can think of something.  But so far no alternatives have borne fruit.

There are also further costs here if we were to go down that route.  For example, a specific 'dictionary' construction syntax would now push us to need a specific 'dictionary' pattern syntax for the deconstruction side of things.  

Finally, the above still gets us to a good place where people can consume this feature in a usable fashion in tons of code.  If it turns out lots of people are writing the pattern you have above and are feeling like it's a problem, we can address that with a future revision where there is some way to "force" dictionary semantics in the case where `List<KVP>` might happen and would not be acceptable.

**Comment by TahirAhmadov** [comment]
I think it would be better to default to dictionary literal if it consists of only spreads and all of them are of type dictionary.

**Reply by jnm2** [reply]
One issue with this is that `foreach (var kvp in [..dict1, ..dict2])` could then iterate in a surprising order or a surprising overall count.

**Reply by TahirAhmadov** [reply]
In that specific use case, wouldn't `dict1.Concat(dict2)` be preferable anyway? 

Also, I'm not sure it's a good idea, but does it make sense to treat the `foreach` expression as a site which expects a target type of `IEnumerable<T>` (in this case `IEnumerable<KVP<K,V>>`)? Forget dictionaries for a moment and think about regular lists, if you have `foreach(var x in [..list1, ..list2]) { ... }`, we don't want to create a new collection and do the copying, and essentially `Concat` behavior is better performance wise, right?

**Reply by jnm2** [reply]
For non-dictionaries, I think I'll be recommending the use of `[..a, ..b]` over the more special case of `a.Concat(b)`. I don't see Concat as particularly preferable. The IDE may end up with a suggestion to replace `a.Concat(b).Append(c)` with `[..a, ..b, c]`.

I think making foreach a special case would be surprising and give the sense that the feature requires coders to keep lists in their head of dictionary-interpretation locations versus collection-interpretation locations for the same syntax. Target typing already makes this a reality, but minimizing this effect from spreading further seems good.

Overall, the aim is to reduce surprise.

**Reply by jnm2** [reply]
A decent counterpoint to the point about surprise is that it could be up to the coder to be sure they disambiguate. If the things being spread are named so that they are obviously dictionaries, like `[..foosByBarId, ..newFoosByBarId]`, then you could guess from context that you're getting a dictionary from that literal. This clarity does hinge on disciplined naming.

**Reply by TahirAhmadov** [reply]
I'm not sure I understand. So, under current working draft, what does `foreach(var x in [..list1, ..list2]) { ... }` compile to?

**Reply by jnm2** [reply]
Looping over a new concatenated collection, with the door open to as much performance optimization as possible to remove actual allocations and copies under the covers. Each element in `list1` in its foreach-enumerated order, followed by each element in `list2` in its.

**Reply by TahirAhmadov** [reply]
OK I see the reasoning here; we want to promise `Concat` type behavior for this type of spread regardless of list vs dict, and regardless of `foreach` or `var` natural type; while reserving the right to optimize inside `foreach` without changing said behavior.

Really, generic inference would alleviate all these concerns, like if I want to combine dicts into one, I can't do `var d3 = [..d1, ..d2];` because `d3` becomes `List<KVP<K,V>>`, and I don't want to type `Dictionary<K, V> d3 = [..d1, ..d2];`, but at least `Dictionary<_, _> d3 = [..d1, ..d2];` would be a little shorter (obviously `K` and `V` would be actual types with potentially long names and it's actually less typing and clearer intent).

**Reply by CyrusNajmabadi** [reply]
> but at least Dictionary<_, _> d3 = [..d1, ..d2]; would be a little shorter (obviously K and V would be actual types with potentially long names and it's actually less typing and clearer intent).

Yes. And this is a space several ldm members want to look into in the nearish future.

**Reply by jnm2** [reply]
`Dictionary<.._>` 😈

**Reply by mungojam** [reply]
> I can't do var d3 = [..d1, ..d2]; because d3 becomes List<KVP<K,V>>, and I don't want to type Dictionary<K, V> d3 = [..d1, ..d2];, but at least Dictionary<_, _> d3 = [..d1, ..d2]; would be a little shorter

I think the proposals for supporting inference from an extension method after the literal like AsDictionary adds reasonable support for var. It's as long as Dictionary<_,_> without forcing you to abandon var if that's your style

**Reply by TahirAhmadov** [reply]
@mungojam actually more like `using static My.Helpers.Collection;   var d3 = Dict([..d1, ..d2]);`

Where `Dict` is `public static Dictionary<K,V> Dict<K,V>(Dictionary<K,V> dict)=>dict;`

The problem here is that this will lead to proliferation of various namespace/class/method names for the same thing. It would be extremely helpful if BCL added blessed versions of these, actually.
